src/parsing.jl

```diff
@@ -0,0 +1,292 @@
+function addkey!(membernames, nam)
+    if !haskey(membernames, nam)
+        membernames[nam] = gensym()
+    end
+    membernames[nam]
+end
+
+onearg(e::Expr, f) = e.head == :call && length(e.args) == 2 && e.args[1] == f
+onearg(e, f) = false
+mapexpr(f, e) = Expr(e.head, map(f, e.args)...)
+
+replace_syms!(x, membernames) = x
+replace_syms!(q::QuoteNode, membernames) =
+    replace_syms!(Meta.quot(q.value), membernames)
+
+function replace_syms!(e::Expr, membernames)
+    if onearg(e, :^)
+        e.args[2]
+    elseif onearg(e, :_I_)
+        @warn "_I_() for escaping variables is deprecated, use cols() instead"
+        addkey!(membernames, :($(e.args[2])))
+    elseif onearg(e, :cols)
+        addkey!(membernames, :($(e.args[2])))
+    elseif e.head == :quote
+        addkey!(membernames, Meta.quot(e.args[1]) )
+    elseif e.head == :.
+        replace_dotted!(e, membernames)
+    else
+        mapexpr(x -> replace_syms!(x, membernames), e)
+    end
+end
+
+protect_replace_syms!(e, membernames) = e
+function protect_replace_syms!(e::Expr, membernames)
+    if e.head == :quote
+        e
+    else
+        replace_syms!(e, membernames)
+    end
+end
+
+function replace_dotted!(e, membernames)
+    x_new = replace_syms!(e.args[1], membernames)
+    y_new = protect_replace_syms!(e.args[2], membernames)
+    Expr(:., x_new, y_new)
+end
+
+
+is_simple_non_broadcast_call(x) = false
+function is_simple_non_broadcast_call(expr::Expr)
+    (expr.head == :call
+        && length(expr.args) >= 2
+        && expr.args[1] isa Symbol
+        && all(x -> x isa QuoteNode || onearg(x, :cols), expr.args[2:end]))
+end
+
+is_simple_broadcast_call(x) = false
+function is_simple_broadcast_call(expr::Expr)
+    (expr.head == :.
+        && length(expr.args) == 2
+        && expr.args[1] isa Symbol
+        && expr.args[2] isa Expr
+        && expr.args[2].head == :tuple
+        && all(x -> x isa QuoteNode || onearg(x, :cols), expr.args[2].args))
+end
+
+function is_simple_function_call(x)
+    is_simple_non_broadcast_call(x) || is_simple_broadcast_call(x)
+end
+
+function args_to_selectors(v)
+    t = map(v) do arg
+        if arg isa QuoteNode
+            arg
+        elseif onearg(arg, :cols)
+            arg.args[2]
+        else
+            Throw(ArgumentError("This path should not be reached, arg: $(arg)"))
```

Comment:
@bkamins @nalimilan 

I would appreciate some feedback on this line. Basically, this function should only get called if all the arguments are either QuoteNodes or `cols` (see `is_simple_broadcast_call`, above). 

But I feel like I'm repeating myself and this error seems heavy-handed. Let me know if there are existing design patterns I should use.

Reply:
can you please give an example, so that I am sure what you want to disallow? Thank you!